File: app/src/main/java/com/gadgetsfolk/kidsgk/fragment/QuizFragment.java
package com.gadgetsfolk.kidsgk.fragment;

import android.content.Intent;
import android.media.MediaPlayer;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.bumptech.glide.Glide;
import com.gadgetsfolk.kidsgk.R;
import com.gadgetsfolk.kidsgk.activity.QuizActivity;
import com.gadgetsfolk.kidsgk.activity.ScoreActivity;
import com.gadgetsfolk.kidsgk.model.Quiz;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.rishabhharit.roundedimageview.RoundedImageView;
import java.util.ArrayList;
import java.util.Collections;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class QuizFragment extends Fragment{
    private int currentPageNumber;
    private Quiz quiz;
    private MediaPlayer mediaCorrect;
    private MediaPlayer mediaWrong;
    @BindView(R.id.img_question)
    RoundedImageView imgQuestion;
    @BindView(R.id.adView)
    AdView mAdView;
    @BindView(R.id.tv_question)
    TextView tvQuestion;
    @BindView(R.id.tv_option_one)
    TextView tvOptionOne;
    @BindView(R.id.tv_option_two)
    TextView tvOptionTwo;
    @BindView(R.id.tv_option_three)
    TextView tvOptionThree;
    @BindView(R.id.tv_option_four)
    TextView tvOptionFour;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_quiz, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getArguments() != null) currentPageNumber = getArguments().getInt("position");

        quiz = QuizActivity.mQuizList.get(currentPageNumber);

        mediaCorrect = MediaPlayer.create(getContext(), R.raw.correct);
        mediaWrong = MediaPlayer.create(getContext(), R.raw.wrong);

        AdRequest adRequest = new AdRequest.Builder().build();
        mAdView.loadAd(adRequest);

        ArrayList<String> options = new ArrayList<>();
        options.add(quiz.getOption_one());
        options.add(quiz.getOption_two());
        options.add(quiz.getOption_three());
        options.add(quiz.getOption_four());

        Collections.shuffle(options);

        tvQuestion.setText(quiz.getQuestion()); //Set Question
        tvOptionOne.setText(options.get(0)); //Set Option One
        tvOptionTwo.setText(options.get(1)); //Set Option Two
        tvOptionThree.setText(options.get(2)); //Set Option Three
        tvOptionFour.setText(options.get(3)); //Set Option Four

        Glide.with(view.getContext())
                .load(quiz.getImg_url())
                .placeholder(R.color.colorGrey)
                .into(imgQuestion);


        mediaWrong.setOnCompletionListener(mediaPlayer -> {
            mediaWrong.release();
            QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        });

        mediaCorrect.setOnCompletionListener(mediaPlayer -> {
            mediaCorrect.release();
            QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        });

    }

    @OnClick(R.id.tv_option_one)
    void onOptionOneSelected(){
        if (tvOptionOne.getText().toString().equals(quiz.getAnswer())) {
            tvOptionOne.setBackgroundResource(R.drawable.bg_answer);
            tvOptionTwo.setOnClickListener(null);
            tvOptionThree.setOnClickListener(null);
            tvOptionFour.setOnClickListener(null);
            mediaCorrect.start();
            QuizActivity.quizScore += 1;
        }else{
            tvOptionOne.setBackgroundResource(R.drawable.bg_wrong_answer);
            tvOptionTwo.setOnClickListener(null);
            tvOptionThree.setOnClickListener(null);
            tvOptionFour.setOnClickListener(null);
            mediaWrong.start();
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }
        setScore(QuizActivity.quizScore);
    }

    @OnClick(R.id.tv_option_two)
    void onOptionTwoSelected(){
        if (tvOptionTwo.getText().toString().equals(quiz.getAnswer())) {
            tvOptionTwo.setBackgroundResource(R.drawable.bg_answer);
            tvOptionThree.setOnClickListener(null);
            tvOptionFour.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            mediaCorrect.start();
            QuizActivity.quizScore += 1;
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }else{
            tvOptionTwo.setBackgroundResource(R.drawable.bg_wrong_answer);
            tvOptionThree.setOnClickListener(null);
            tvOptionFour.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            mediaWrong.start();
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }
        setScore(QuizActivity.quizScore);
    }

    @OnClick(R.id.tv_option_three)
    void onOptionThreeSelected(){
        if (tvOptionThree.getText().toString().equals(quiz.getAnswer())) {
            tvOptionThree.setBackgroundResource(R.drawable.bg_answer);
            tvOptionFour.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            tvOptionTwo.setOnClickListener(null);
            mediaCorrect.start();
            QuizActivity.quizScore += 1;
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }else{
            tvOptionThree.setBackgroundResource(R.drawable.bg_wrong_answer);
            tvOptionFour.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            tvOptionTwo.setOnClickListener(null);
            mediaWrong.start();
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }
        setScore(QuizActivity.quizScore);
    }

    @OnClick(R.id.tv_option_four)
    void onOptionFourSelected(){
        if (tvOptionFour.getText().toString().equals(quiz.getAnswer())) {
            tvOptionFour.setBackgroundResource(R.drawable.bg_answer);
            tvOptionThree.setOnClickListener(null);
            tvOptionTwo.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            mediaCorrect.start();
            QuizActivity.quizScore += 1;
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }else{
            tvOptionFour.setBackgroundResource(R.drawable.bg_wrong_answer);
            tvOptionThree.setOnClickListener(null);
            tvOptionTwo.setOnClickListener(null);
            tvOptionOne.setOnClickListener(null);
            mediaWrong.start();
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }
        setScore(QuizActivity.quizScore);
    }

    private void setScore(int score){
        QuizActivity.questionsAttempted += 1;
        //Log.e("att", String.valueOf(QuizActivity.questionsAttempted));
        if (QuizActivity.questionsAttempted == QuizActivity.mQuizList.size()){
            QuizActivity.score.setCorrect_questions(String.valueOf(score));
            QuizActivity.score.setTotal_questions(String.valueOf(QuizActivity.mQuizList.size()));

            int percentage = (QuizActivity.quizScore / Integer.parseInt(QuizActivity.score.getTotal_questions())) * 100;
            QuizActivity.score.setPercentage(String.valueOf(percentage)); //Set Percentage
            //Give remark according to percentage
            if (percentage <= 100 && percentage > 75)
                QuizActivity.score.setRemark("Excellent!");
            else if (percentage <= 75 && percentage > 60)
                QuizActivity.score.setRemark("Can do better");
            else if (percentage <= 60 && percentage > 50)
                QuizActivity.score.setRemark("Keep it up");
            else if (percentage <= 50 && percentage > 40)
                QuizActivity.score.setRemark("Practice more");
            else
                QuizActivity.score.setRemark("Learn and play more quizzes");

            Intent intent = new Intent(getContext(), ScoreActivity.class);
            intent.putExtra("score_obj", QuizActivity.score);
            startActivity(intent);
            if (getActivity() != null) getActivity().finish();
        }
    }
}

File: app/src/main/java/com/gadgetsfolk/kidsgk/fragment/QuizFragment.kt
package com.gadgetsfolk.kidsgk.fragment

import android.content.Intent
import android.media.MediaPlayer
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.bumptech.glide.Glide
import com.gadgetsfolk.kidsgk.R
import com.gadgetsfolk.kidsgk.activity.QuizActivity
import com.gadgetsfolk.kidsgk.activity.ScoreActivity
import com.gadgetsfolk.kidsgk.databinding.FragmentQuizBinding
import com.gadgetsfolk.kidsgk.model.Quiz
import com.google.android.gms.ads.AdRequest
import java.util.*

class QuizFragment : Fragment() {
    private var currentPageNumber = 0
    private var quiz: Quiz? = null
    private var mediaCorrect: MediaPlayer? = null
    private var mediaWrong: MediaPlayer? = null

    private var _binding: FragmentQuizBinding? = null
    //This property is only valid between onCreateView and
    //onDestroyView
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentQuizBinding.inflate(inflater, container, false)
        return  binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (arguments != null) currentPageNumber = arguments!!.getInt("position")
        quiz = QuizActivity.mQuizList!![currentPageNumber]
        mediaCorrect = MediaPlayer.create(context, R.raw.correct)
        mediaWrong = MediaPlayer.create(context, R.raw.wrong)
        val adRequest = AdRequest.Builder().build()
        binding.adView.loadAd(adRequest)
        val options = ArrayList<String?>()
        options.add(quiz!!.option_one)
        options.add(quiz!!.option_two)
        options.add(quiz!!.option_three)
        options.add(quiz!!.option_four)
        options.shuffle()
        binding.tvQuestion.text = quiz!!.question //Set Question
        binding.tvOptionOne.text = options[0] //Set Option One
        binding.tvOptionTwo.text = options[1] //Set Option Two
        binding.tvOptionThree.text = options[2] //Set Option Three
        binding.tvOptionFour.text = options[3] //Set Option Four

        Glide.with(view.context).load(quiz!!.img_url).placeholder(R.color.colorGrey).into(binding.imgQuestion)

        mediaWrong?.setOnCompletionListener {
            mediaWrong?.release()
            QuizActivity.nonViewPager?.currentItem = currentPageNumber + 1
        }

        mediaCorrect?.setOnCompletionListener {
            mediaCorrect?.release()
            QuizActivity.nonViewPager?.currentItem = currentPageNumber + 1
        }

        binding.tvOptionOne.setOnClickListener{ onOptionOneSelected() }
        binding.tvOptionTwo.setOnClickListener{ onOptionTwoSelected() }
        binding.tvOptionThree.setOnClickListener{ onOptionThreeSelected() }
        binding.tvOptionFour.setOnClickListener{ onOptionFourSelected() }

    }

    private fun onOptionOneSelected() {
        if (binding.tvOptionOne.text.toString() == quiz!!.answer) {
            binding.tvOptionOne.setBackgroundResource(R.drawable.bg_answer)
            binding.tvOptionTwo.setOnClickListener(null)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionFour.setOnClickListener(null)
            mediaCorrect!!.start()
            QuizActivity.quizScore += 1
        } else {
            binding.tvOptionOne.setBackgroundResource(R.drawable.bg_wrong_answer)
            binding.tvOptionTwo.setOnClickListener(null)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionFour.setOnClickListener(null)
            mediaWrong!!.start()
            //QuizActivity.viewPager.setCurrentItem(currentPageNumber + 1);
        }
        setScore(QuizActivity.quizScore)
    }

    private fun onOptionTwoSelected() {
        if (binding.tvOptionTwo.text.toString() == quiz!!.answer) {
            binding.tvOptionTwo.setBackgroundResource(R.drawable.bg_answer)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionFour.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            mediaCorrect!!.start()
            QuizActivity.quizScore += 1
        } else {
            binding.tvOptionTwo.setBackgroundResource(R.drawable.bg_wrong_answer)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionFour.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            mediaWrong!!.start()
        }
        setScore(QuizActivity.quizScore)
    }

    private fun onOptionThreeSelected() {
        if (binding.tvOptionThree.text.toString() == quiz!!.answer) {
            binding.tvOptionThree.setBackgroundResource(R.drawable.bg_answer)
            binding.tvOptionFour.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            binding.tvOptionTwo.setOnClickListener(null)
            mediaCorrect!!.start()
            QuizActivity.quizScore += 1
        } else {
            binding.tvOptionThree.setBackgroundResource(R.drawable.bg_wrong_answer)
            binding.tvOptionFour.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            binding.tvOptionTwo.setOnClickListener(null)
            mediaWrong!!.start()
        }
        setScore(QuizActivity.quizScore)
    }

    private fun onOptionFourSelected() {
        if (binding.tvOptionFour.text.toString() == quiz!!.answer) {
            binding.tvOptionFour.setBackgroundResource(R.drawable.bg_answer)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionTwo.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            mediaCorrect!!.start()
            QuizActivity.quizScore += 1
        } else {
            binding.tvOptionFour.setBackgroundResource(R.drawable.bg_wrong_answer)
            binding.tvOptionThree.setOnClickListener(null)
            binding.tvOptionTwo.setOnClickListener(null)
            binding.tvOptionOne.setOnClickListener(null)
            mediaWrong!!.start()
        }
        setScore(QuizActivity.quizScore)
    }

    private fun setScore(score: Int) {
        QuizActivity.questionsAttempted += 1
        if (QuizActivity.questionsAttempted == QuizActivity.mQuizList!!.size) {
            QuizActivity.score.correct_questions = score.toString()
            QuizActivity.score.total_questions = QuizActivity.mQuizList!!.size.toString()
            val percentage = QuizActivity.quizScore / QuizActivity.score.total_questions!!.toInt() * 100
            QuizActivity.score.percentage = percentage.toString() //Set Percentage
            //Give remark according to percentage
            when (percentage) {
                in 76..100 -> QuizActivity.score.remark = "Excellent!"
                in 61..75 -> QuizActivity.score.remark = "Can do better"
                in 51..60 -> QuizActivity.score.remark = "Keep it up"
                in 41..50 -> QuizActivity.score.remark = "Practice more"
                else -> QuizActivity.score.remark = "Learn and play more quizzes"
            }
            val intent = Intent(context, ScoreActivity::class.java)
            intent.putExtra("score_obj", QuizActivity.score)
            startActivity(intent)
            if (activity != null) activity!!.finish()
        }
    }
}